src/Fantomas/TokenParser.fs

```diff
@@ -90,91 +90,189 @@ let private createHashToken lineNumber content fullMatchedLength offset =
-
-    let findDefineInLine idx line =
-        let lineNumber = idx + 1
-        let fullMatchedLength = String.length line
-        let trimmed = line.TrimStart()
-
-        let offset =
-            String.length line - String.length trimmed
-
-        if trimmed.StartsWith("#if")
-        then processLine "#if" trimmed lineNumber fullMatchedLength offset
-        elif trimmed.StartsWith("#elseif")
-        then processLine "#elseif" trimmed lineNumber fullMatchedLength offset
-        elif trimmed.StartsWith("#else")
-        then processLine "#else" trimmed lineNumber fullMatchedLength offset
-        elif trimmed.StartsWith("#endif")
-        then processLine "#endif" trimmed lineNumber fullMatchedLength offset
-        else []
-
-    let hasUnEvenAmount regex line =
-        (Regex.Matches(line, regex).Count
-         - Regex.Matches(line, "\\\\" + regex).Count) % 2 = 1
-
-    let singleQuoteWrappedInTriple line =
-        Regex
-            .Match(line, "\\\"\\\"\\\".*\\\".*\\\"\\\"\\\"")
-            .Success
-
-    sourceCode
-    |> String.normalizeThenSplitNewLine
-    |> Array.indexed
-    |> Array.fold (fun (state, defines) (idx, line) ->
-        let hasTripleQuotes = hasUnEvenAmount "\"\"\"" line
-
-        let hasSingleQuote =
-            (not hasTripleQuotes)
-            && (hasUnEvenAmount "\"" line)
-            && not (singleQuoteWrappedInTriple line)
-
-        match state with
-        | Normal when (hasTripleQuotes) -> InsideTripleQuoteString, defines
-        | Normal when (hasSingleQuote) -> InsideSingleQuoteString, defines
-        | Normal when (not hasTripleQuotes && not hasSingleQuote) -> Normal, (findDefineInLine idx line) :: defines
-
-        | InsideTripleQuoteString when (not hasTripleQuotes) -> InsideTripleQuoteString, defines
-        | InsideTripleQuoteString when hasTripleQuotes -> Normal, defines
-
-        | InsideSingleQuoteString when (not hasSingleQuote) -> InsideSingleQuoteString, defines
-        | InsideSingleQuoteString when (hasSingleQuote) -> Normal, defines
-
-        | _ -> failwithf "unexpected %A" state
-
-        ) (Normal, [])
-    |> snd
-    |> List.rev
-    |> List.concat
-
-and tokenize defines (content: string): Token list =
+            tokens
+            |> List.map (fun t ->
+                let info =
+                    { t.TokenInfo with
+                          LeftColumn = t.TokenInfo.LeftColumn + hashContentLength
+                          RightColumn = t.TokenInfo.RightColumn + hashContentLength }
+
+                { t with
+                      LineNumber = lineNumber
+                      TokenInfo = info })
+            |> fun rest ->
+                (createHashToken lineNumber hashContent offset)
+                :: rest
+
+        let sourceLength = String.length sourceCode
+        let lastIndex = sourceLength - 3
+        let defines = ResizeArray()
+        let mutable currentState = Normal
+        let mutable newlineIndexes = ResizeArray()
```

Comment:
Pretty sure this won't need to be `mutable` (since the `ResizeArray` is already mutable).

If this were code in G-Research, I'd comment that this should be expressed immutably as a fold or by recursion; iterating over a list updating a state machine is a very suitable task for `fold`. But if imperative style is normal in Fantomas, then fine.

Reply:
I went with an imperative style for no good reason. Refactoring to a `fold` was the right choice here.